segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,500 @@
+    private final ScheduledExecutorService executor;
+
+    //region Constructor
+
+    ContainerEventProcessorImpl(@NonNull SegmentContainer container, @NonNull MetadataStore metadataStore,
+                                @NonNull Duration iterationDelay, @NonNull Duration containerOperationTimeout,
+                                @NonNull ScheduledExecutorService executor) {
+        this(container.getId(), getOrCreateInternalSegment(container, metadataStore, containerOperationTimeout), iterationDelay,
+                containerOperationTimeout, executor);
+    }
+
+    @VisibleForTesting
+    ContainerEventProcessorImpl(int containerId, @NonNull Function<String, CompletableFuture<DirectSegmentAccess>> segmentSupplier,
+                                @NonNull Duration iterationDelay, @NonNull Duration containerOperationTimeout,
+                                @NonNull ScheduledExecutorService executor) {
+        this.containerId = containerId;
+        this.traceObjectId = String.format("ContainerEventProcessor[%d]", containerId);
+        this.segmentSupplier = segmentSupplier;
+        this.iterationDelay = iterationDelay;
+        this.containerOperationTimeout = containerOperationTimeout;
+        this.closed = new AtomicBoolean(false);
+        this.executor = executor;
+        // This class just reports the metrics for all the registered EventProcessor objects.
+        this.eventProcessorMetricsReporting = executor.scheduleAtFixedRate(this::reportMetrics, iterationDelay.toMillis(),
+                iterationDelay.toMillis(), TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * This function instantiates a new Segment supplier by creating actual Segment against the {@link SegmentContainer}
+     * passed as input based on the name passed to the function. Note that the Segment is only created if it does not
+     * exists. Otherwise, it is just loaded and returned.
+     *
+     * @param container {@link SegmentContainer} to create Segments from.
+     * @param timeout   Timeout for the Segment creation to complete.
+     * @return A future that, when completed, contains reference to the Segment to be used by a given
+     * {@link ContainerEventProcessor.EventProcessor} based on its name.
+     */
+    private static Function<String, CompletableFuture<DirectSegmentAccess>> getOrCreateInternalSegment(SegmentContainer container,
+                                                                                                       MetadataStore metadataStore,
+                                                                                                       Duration timeout) {
+        return s -> Futures.exceptionallyComposeExpecting(
+                container.forSegment(getEventProcessorSegmentName(container.getId(), s), timeout),
+                e -> e instanceof StreamSegmentNotExistsException,
+                () -> metadataStore.submitAssignment(newSegment(getEventProcessorSegmentName(container.getId(), s),
```

Comment:
Hmm, I wonder why `submitAssignment` is visible here? Was it `VisibleForTesting`? 

I am not very keen on invoking this internal method from outside that class. Could you verify why `submitAssignment` was visible here in the first place? If only for testing, I strongly recommend that we make another method on `MetadataStore` (like `registerPinnedSegment` or alike) where we hide some of these complexities. It should only take a segment name, type and attributes.

Reply:
The visibility of this method is `protected`, and given that both classes are under the same package, calling it is legal. Also, this method had no `@VisibleForTesting` annotation and it was properly documented. Given that, I think that using it is fine and if we want to refactor the `MetadataStore` class and how other classes access it, we can do it separately.